tests/llmobs/test_llmobs.py

```diff
@@ -98,6 +102,157 @@ def test_propagates_ignore_non_llmobs_spans(self, tracer, llmobs_events):
+    def test_annotated_span(self, llmobs, llmobs_enable_opts, llmobs_events):
+        with llmobs.annotation_context(tags={"scrub_values": "1"}):
+            with llmobs.llm() as llm_span:
+                llmobs.annotate(llm_span, input_data="value", output_data="value")
+        assert llmobs_events[0]["meta"]["input"] == {"messages": [{"content": "redacted", "role": ""}]}
+        assert llmobs_events[0]["meta"]["output"] == {"messages": [{"content": "redacted", "role": ""}]}
+
+    def _bad_return_type_processor(span: LLMObsSpan) -> str:
+        return "not a span"
+
+    @pytest.mark.parametrize("llmobs_enable_opts", [dict(span_processor=_bad_return_type_processor)])
+    def test_processor_bad_return_type(self, llmobs, llmobs_events):
+        """Test that a processor that returns a non-LLMObsSpan type is ignored."""
+        with llmobs.llm() as llm_span:
+            llmobs.annotate(llm_span, input_data="value", output_data="value")
+        assert llmobs_events[0]["meta"]["input"] == {"messages": [{"content": "value", "role": ""}]}
+        assert llmobs_events[0]["meta"]["output"] == {"messages": [{"content": "value", "role": ""}]}
+
+    def test_ddtrace_run_register_processor(self, ddtrace_run_python_code_in_subprocess, llmobs_backend):
+        """Users using ddtrace-run can register a processor to be called on each LLMObs span."""
+        env = os.environ.copy()
+        env["DD_LLMOBS_ML_APP"] = "test-ml-app"
+        env["DD_API_KEY"] = "test-api-key"
+        env["DD_LLMOBS_ENABLED"] = "1"
+        env["DD_LLMOBS_AGENTLESS_ENABLED"] = "0"
+        env["DD_TRACE_ENABLED"] = "0"
+        env["DD_TRACE_AGENT_URL"] = llmobs_backend.url()
+        out, err, status, _ = ddtrace_run_python_code_in_subprocess(
+            dedent(
+                """
+            from ddtrace.llmobs import LLMObs
+
+            def span_processor(s):
+                if s.get_tag("scrub_values") == "1":
+                    s.input_messages = [{"content": "scrubbed"}]
+                    s.output_messages = [{"content": "scrubbed"}]
+                return s
+
+            LLMObs.register_processor(span_processor)
+
+            with LLMObs.llm("openai.request") as llm_span:
+                LLMObs.annotate(llm_span, input_data="value", output_data="value", tags={"scrub_values": "1"})
+            with LLMObs.llm("openai.request") as llm_span:
+                LLMObs.annotate(llm_span, input_data="value", output_data="value", tags={"scrub_values": "0"})
+            """
+            ),
+            env=env,
+        )
+        assert out == b""
+        assert status == 0, err
+        assert err.decode() == ""
+        events = llmobs_backend.wait_for_num_events(num=1)
+        traces = events[0]
+        assert len(traces) == 2
+        assert "scrub_values:1" in traces[0]["spans"][0]["tags"]
+        assert traces[0]["spans"][0]["meta"]["input"]["messages"][0]["content"] == "scrubbed"
```

Comment:
gonna follow up to make fetching the spans / traces from events less implementation dependent